### ISSUE TITLE
Add support for attaching event listeners to the global window

### ISSUE THREAD

**Issue by vultix** [DioxusLabs/dioxus#2405]
## Feature Request

It's common to want to attach event listeners to the global window. For example, a custom slider element will want to listen to mousedown locally on the slider, but mousemove globally on the window.

Svelte has a [`svelte:window`](https://learn.svelte.dev/tutorial/svelte-window) special element for exactly these use cases.

## Implement Suggestion

Taking a cue from svelte, we could add something like a window special element that events can be attached to:
```rust
rsx! {
    dioxus:window {
        onmousemove: move |event| handle_mousemove(event)
    }
    div {
        id: "slider",
        onmousedown: move |event| handle_mousedown(event),
        ...
    }
}
```

We could also introduce a new `use_window_event` hook or something similar.

**Comment by DogeDark** [comment]
On the desktop platform there is ``use_wry_event_handler``. We use it in SDK [here](https://github.com/DioxusLabs/sdk/blob/57ab3fa972c6d4a7acc25e82a0aafc3ff9e63403/sdk/src/utils/window.rs#L91). I don't believe anything similar exists for Web though, aside from using the ``web-sys`` crate.

**Comment by ealmloff** [comment]
There is some overlap with https://github.com/DioxusLabs/dioxus/issues/628, and https://github.com/DioxusLabs/dioxus/issues/1169. Some way to attach attributes/listeners to somewhere outside of children in Dioxus unlocks a bunch of new usecases

**Comment by poly2it** [comment]
Any progress? Really in need of this.

**Comment by dignifiedquire** [comment]
based on other issues and discussions I have written this

https://gist.github.com/dignifiedquire/eca103b7f93604dd367efce8a59e0218

which allows you to write

```rust
use_on_window("keydown", |evt: web_sys::KeyboardEvent| {
    tracing::info!("keydown: {:?}, {:?}", evt, evt.key());
});
```

**Comment by poly2it** [comment]
> based on other issues and discussions I have written this
> 
> https://gist.github.com/dignifiedquire/eca103b7f93604dd367efce8a59e0218
> 
> which allows you to write
> 
> use_on_window("keydown", |evt: web_sys::KeyboardEvent| {
>     tracing::info!("keydown: {:?}, {:?}", evt, evt.key());
> });

This worked really well, but breaks in fullstack. Do you know how I can avoid errors like `cannot access imported statics on non-wasm targets` when using this in fullstack?

**Comment by ealmloff** [comment]
> > based on other issues and discussions I have written this
> > https://gist.github.com/dignifiedquire/eca103b7f93604dd367efce8a59e0218
> > which allows you to write
> > use_on_window("keydown", |evt: web_sys::KeyboardEvent| {
> > tracing::info!("keydown: {:?}, {:?}", evt, evt.key());
> > });
> 
> This worked really well, but breaks in fullstack. Do you know how I can avoid errors like `cannot access imported statics on non-wasm targets` when using this in fullstack?

Try moving the client-side logic behind a [`client!`](https://docs.rs/dioxus/latest/dioxus/prelude/macro.client.html) block which will only compile if the client side feature is enabled

**Comment by poly2it** [comment]
> > > based on other issues and discussions I have written this
> > > https://gist.github.com/dignifiedquire/eca103b7f93604dd367efce8a59e0218
> > > which allows you to write
> > > use_on_window("keydown", |evt: web_sys::KeyboardEvent| {
> > > tracing::info!("keydown: {:?}, {:?}", evt, evt.key());
> > > });
> > 
> > 
> > This worked really well, but breaks in fullstack. Do you know how I can avoid errors like `cannot access imported statics on non-wasm targets` when using this in fullstack?
> 
> Try moving the client-side logic behind a [`client!`](https://docs.rs/dioxus/latest/dioxus/prelude/macro.client.html) block which will only compile if the client side feature is enabled

I tried. Sadly I get a runtime error on the server:
```
...
18:09:55 [server]   19: <once_cell::unsync::Lazy<T,F> as core::ops::deref::Deref>::deref
18:09:55 [server]              at /home/holst/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/once_cell-1.21.3/src/lib.rs:843:13
18:09:55 [server]   20: js_sys::global
18:09:55 [server]              at /home/holst/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/js-sys-0.3.77/src/lib.rs:6051:12
18:09:55 [server]   21: web_sys::window
18:09:55 [server]              at /home/holst/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/web-sys-0.3.77/src/lib.rs:38:5
18:09:55 [server]   22: kiosk::window_events::use_on_window
18:09:55 [server]              at ./src/window_events.rs:17:15
18:09:55 [server]   23: kiosk::components::background_dots::BackgroundDots
18:09:55 [server]              at ./src/components/background_dots.rs:12:3
18:09:55 [server]   24: core::ops::function::Fn::call
18:09:55 [server]              at /rustc/7b84c9e9ca4b9c68c888b77762c31e180ee7af45/library/core/src/ops/function.rs:79:5
...
```

You can check it out here: https://github.com/poly2it/kiosk/blob/2864d0cbdb0015dff9cac5e3d94e9697bc9e5a73/src/components/background_dots.rs#L10.

**Comment by ealmloff** [comment]
It looks like you have the web feature enabled by default [here](https://github.com/poly2it/kiosk/blob/2864d0cbdb0015dff9cac5e3d94e9697bc9e5a73/Cargo.toml#L20) that could be causing the block to run even when the server feature is enabled

**Comment by poly2it** [comment]
> It looks like you have the web feature enabled by default [here](https://github.com/poly2it/kiosk/blob/2864d0cbdb0015dff9cac5e3d94e9697bc9e5a73/Cargo.toml#L20) that could be causing the block to run even when the server feature is enabled

Indeed, that was it. Thanks for pointing it out. It had been left that way since I migrated to fullstack.

**Comment by onfranciis** [comment]
> based on other issues and discussions I have written this
> 
> https://gist.github.com/dignifiedquire/eca103b7f93604dd367efce8a59e0218
> 
> which allows you to write
> 
> use_on_window("keydown", |evt: web_sys::KeyboardEvent| {
>     tracing::info!("keydown: {:?}, {:?}", evt, evt.key());
> });

It worked 🥹
I've given you a follow. Thank you
